torch/csrc/jit/tensorexpr/cuda_codegen.cpp

```diff
@@ -661,6 +663,148 @@ class PrioritizeLoad : public IRMutator {
   std::unordered_set<const Var*> thread_local_bufs_;
 };
 
+class CudaDependencyHandler : public IRMutator {
+ public:
+  CudaDependencyHandler(analysis::MemDependencyChecker* checker)
+      : mem_dependency_checker_(checker) {}
+
+  Stmt* mutate(const Block* v) override {
+    auto stmts = v->stmts();
+    auto newStmts = handleDependences({stmts.begin(), stmts.end()});
+    return new Block(newStmts);
+  }
+
+ private:
+  // Check if the given statements have the same thread index.
+  bool sameThreadIdx(Stmt* s1, Stmt* s2) {
+    auto enclosing_for = [](Stmt* s) -> For* {
+      Stmt* curr = s;
+      while (curr) {
+        if (auto f = dynamic_cast<For*>(curr)) {
+          return f;
+        }
+        curr = curr->get_parent();
+      }
+      return nullptr;
+    };
+    auto for1 = enclosing_for(s1);
+    auto for2 = enclosing_for(s2);
+    if (for1 && for2) {
+      auto loopOptions1 = for1->loop_options();
+      auto loopOptions2 = for2->loop_options();
+      if (loopOptions1.is_gpu_thread_index() &&
+          loopOptions2.is_gpu_thread_index()) {
+        return loopOptions1.gpu_thread_index() ==
+            loopOptions2.gpu_thread_index();
+      }
+    }
+    return false;
+  }
+
+  // Identify the list of buffers that were written to in `s1` and read in `s2`
+  // and hence need synchronization before `s2` is executed.
+  std::unordered_set<const Buf*> buffersNeedingSync(Stmt* s1, Stmt* s2) {
+    auto s1Bounds = getInferredBounds(*mem_dependency_checker_, s1, true);
+    auto s2Bounds = getInferredBounds(*mem_dependency_checker_, s2, true);
+    std::unordered_set<const Buf*> syncBuffers;
+    for (const auto& s2Bound : s2Bounds) {
+      const Buf* buf = s2Bound.first;
+      if (s1Bounds.find(buf) == s1Bounds.end()) {
+        continue;
+      }
+
+      // We only handle RAW hazards now.
```

Comment:
What's the reasoning behind this limitation?

Reply:
The reason is the logic we are using to figure out if a buffer has been "synchronized" at a given point. Once we insert a `__syncthreads`, we mark all buffers written so far as "synchronized". 

For WAW or WAR hazards, the buffers will be written more than once. That means that we can not associate `__syncthreads` with a buffer. 

We can handle those by associating `__syncthreads` with `Store`s instead I think. But I wanted to get a simple case working correctly first. I can handle the rest of the cases as follow ups.